Add shared `useBeforeUnloadPromptEffect`

tomalec opened the pull request:
### Changes proposed in this Pull Request:

Add shared `useBeforeUnloadPromptEffect`.
Use it in edit flow.
Make the effect prompt the user also when `woocommerce/navigation` is used to navigate, not only page reloads.

Needed for: https://github.com/woocommerce/google-listings-and-ads/issues/156

### Screenshots:


https://user-images.githubusercontent.com/17435/112763686-c5a39380-9005-11eb-9ecc-5f7ee2619a96.mp4


### Detailed test instructions:

##### Edit flow
1. Go to **[Marketing > GLA > Dashboard](https://gla1.test/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fdashboard) > [Edit](https://gla1.test/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fedit-free-campaign&programId=123)**
2. Do some changes to the form
2. Navigate between steps, by clicking on step titles in the top bar (no prompt should be shown)
3. Navigate away using:
	1. Using `woocommerce/navigation` bound / SPA action, like `<` button in the topbar. (Prompt with the custom message should be shown)
	2. Using a regular browser reload: refresh/back/forward button, a static link on the site.

~Please note, that currently the prompt will be shown, even if no changes to the for were made, this will be covered in a separate PR.~

##### Setup
1. Go to **[Marketing > GLA > Dashboard > Add paid campaign](https://gla1.test/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fsetup-ads)**
2. Do some changes to the form
3. Navigate away using:
	1. Using `woocommerce/navigation` bound / SPA action, like `<` button in the topbar. (Prompt with the custom message should be shown)
	2. Using a regular browser reload: refresh/back/forward button, a static link on the site.


### Changelog Note:

> Prompt leaving the form with any kind of `woocommerce/navigation` away.

### Doubts:

1. Ths PR uses API of [`history` npm package](https://github.com/ReactTraining/history/). The `history` object is exposed by `woocommerce/navigation#getHistory()`, but is not very documented. https://www.npmjs.com/package/@woocommerce/navigation?activeTab=readme#gethistory So there is a risk of this being considered an implementation detail and being changed.
2. We are using the out-dated ([v4](https://github.com/ReactTraining/history/blob/v4/docs/Blocking.md)) API of `history` package (current [v5](https://github.com/ReactTraining/history/blob/28c89f4091ae9e1b0001341ea60c629674e83627/docs/blocking-transitions.md)). v5 API seems more flexible, and would do more work for us. I guess bumping the version in `woocommerce/navigation` could introduce a large backward-compat issue, for other plugins. Given the point above, not to lock us harder on technical debt, I was thinking of making us future-proof and support both APIs, but that may be overkill.


### Additional notes:

1. Previously we were binding prompt for leaving the page manually, like in https://github.com/woocommerce/google-listings-and-ads/blob/feature/156-save-edited-campaign/js/src/setup-ads/top-bar/index.js#L18-L34. That worked with Setup, when we have limited number of actions to navigate away. But, I'd say it will not scale or be future-proof, when we may add a new link and forget about binding the prompt. This PR, binds to any `woocommerce/-` or browser navigations
2. :warning: With the previous implementation, the tracking event was not recorded if the navigation was stopped. With the new one, we record every, even canceled intent to go back. With the new approach, it could be tricky to bind after the `getHistory().block`. //cc @jconroy 
[`feature/156-promptEffect-2?expand=1`#diff-3417a890fd](https://github.com/woocommerce/google-listings-and-ads/compare/feature/156-promptEffect-2?expand=1#diff-3417a890fdbbdb5379d6880ecae871a41fbf7a0bed33c0f37896d4e8d30f1267)L19-L33

tomalec committed:
Add shared `useBeforeUnloadPromptEffect`.

Use it in edit flow.

(cherry picked from commit 60c1f0e642b558e85d8ad15fc8fd470df61fffca)

tomalec committed:
Use `getHistory().block` API, to confirm navigating away.

(cherry picked from commit 400121f0fe188ece9dd93b719572dd1e6028831a)

ecgan reviewed (changes requested):
Just two things that I would prefer to change: 

1. Use set object instead of an array.
2. Change the naming and signature of `useBeforeUnloadPromptEffect`.

Other than that, I think it's pretty good, especially with the comments 😄 👍

tomalec committed:
Fix spelling in .~/hooks/useBeforeUnloadPromptEffect.js code docs.

Co-authored-by: Gan Eng Chin <ecgan@users.noreply.github.com>

tomalec committed:
Use `Set` for allow list of locations in `EditFreeCampaign`.

Addresses https://github.com/woocommerce/google-listings-and-ads/pull/393#discussion_r603295629.

tomalec committed:
Rename useBeforeUnloadPromptEffect to useNavigateAwayPromptEffect.

Addresses: https://github.com/woocommerce/google-listings-and-ads/pull/393#discussion_r603329751.

tomalec committed:
Make message a required param for useNavigateAwayPromptEffect.

Addresses: https://github.com/woocommerce/google-listings-and-ads/pull/393#discussion_r603993356.

tomalec committed:
Add the default step to the list of allowed navigations from EditFreeCampaign.

Addresses: https://github.com/woocommerce/google-listings-and-ads/pull/393#discussion_r603978257.

tomalec committed:
Merge branch 'trunk' into feature/156-promptEffect-2

Conflicts:
	js/src/edit-free-campaign/index.js

tomalec committed:
Add dirty check to EditFreeCampaign

(cherry picked from commit 400121f0fe188ece9dd93b719572dd1e6028831a)

eason9487 reviewed (approved):
Apart from the changes of event tracking are awaiting the decision, the rest parts look good to me.

tomalec committed:
Rearrange params for useNavigateAwayPromptEffect,

as suggested at https://github.com/woocommerce/google-listings-and-ads/pull/393#discussion_r604691446.
